### ISSUE TITLE
using sparse conv is slower

### ISSUE THREAD

**Issue by IwakuraRein** [dabeschte/DeltaCNN#7]
There is a module inside my network:

```python
self.network = nn.Sequential(nn.Conv2d(7, 32, 3, padding=1),
                             nn.ReLU(),
                             nn.Conv2d(32, 32, 3, padding=1),
                             nn.ReLU(),
                             nn.Conv2d(32, 32, 3, padding=1),
                             nn.ReLU(),
                             nn.Conv2d(32, 14, 3, padding=1),
                             nn.ReLU()
                             )
```

I replaced it with sparse operations:

```python
self.sparsify = dc.DCSparsify(delta_threshold=0.1, dilation=5)
self.densify = dc.DCDensify()
self.network = nn.Sequential(dc.DCConv2d(7, 32, 3, padding=1),
                             dc.DCActivation(activation="relu"),
                             dc.DCConv2d(32, 32, 3, padding=1),
                             dc.DCActivation(activation="relu"),
                             dc.DCConv2d(32, 32, 3, padding=1),
                             dc.DCActivation(activation="relu"),
                             dc.DCConv2d(32, 14, 3, padding=1),
                             dc.DCActivation(activation="relu")
                             )
```

Then I tested its time cost on a Titan RTX GPU:

```python
# ...
def forward(self, inp):
    torch.cuda.synchronize()
    starter, ender = torch.cuda.Event(enable_timing=True), torch.cuda.Event(enable_timing=True)
    starter.record()

    encodes = self.densify(self.network(self.sparsify(inp)))
    # encodes = self.network(inp)

    ender.record()
    torch.cuda.synchronize()
    print(starter.elapsed_time(ender))
    # ...
```

The original time cost is 8.2-8.5ms. However, the time cost after implementing sparse operations increased to 15ms. I am pretty sure the input frames have huge coherence. How can I fix this?

**Comment by dabeschte** [comment]
Hey!
That's interesting. Especially since the 3x3 kernels should be well optimized. But I think the main issue here is that the feature maps are quite small (few channels). In the current implementation, a convolution with a single output channel will take roughly as long as a convolution with 64 output channels due to the way the convolution is parallelized.

I added support for 3x3 convolutions with 32 output channels that should give you a small speedup over previous version, but don't expect too much of a difference with small batch sizes. To try it out, pull the latest commit and run the setup again.

On my PC (1080 ti), deltacnn vs cudnn performance is as follows with an input of size (batch_size, 7, 128, 128) (10 runs, taking fastest run):
- batch size 1: cudnn: 0.90ms, deltacnn w/o sparsity (negative threshold): 1.15ms, deltacnn fully sparse  (inf threshold): 0.89ms
- batch size 128: cudnn: 58.0ms, deltacnn w/o sparsity (negative threshold): 60.4ms, deltacnn fully sparse (inf threshold): 5.3ms

When you use layer with more channels, deltacnn can sometimes even outperform cudnn in dense mode. for example with 128 channels (instead of 32) and a batch size of 1: cudnn 3.63ms, deltacnn w/o sparsity 3.17ms.

Let me know if that answers your question.

**Comment by IwakuraRein** [comment]
@dabeschte ，
Thanks for your response. The input size in my previous experiment is {1, 7, 720, 1280} (i.e., I am working on a 720p video stream). Does larger batch size improve DeltaCNN's performance? Maybe I should slice the 720*1280 input into small patches.

**Comment by IwakuraRein** [comment]
I tested the random input with size {128, 90, 80, 7} and it got slower. The inference time was about 17ms.

**Comment by dabeschte** [comment]
When using such a large resolution, a large batch size is not necessary. I am surprised that the performance is so bad on your PC.
Still, if there is enough sparsity in the image, DeltaCNN should still be able to outperform cuDNN.
For a 3x3 convolution, DeltaCNN uses an 8x8 input per tile and processes a 6x6 output. Speedup is gained when all of the 8x8 input pixels remain the same as before (i.e., when the mask of all of those pixels is set to 0).
With a random input (assuming torch.rand((1,7,720,1280))) and a threshold of 0.1, most like all tiles will require an update, probably even every single pixel.

I'd suggest a few steps to increase sparsity:
- verify feature map sparsity: density = self.sparsify(inp)[1].float().mean()
- check if the updates (the "ones") (in  self.sparsify(inp)[1]) are spread evenly over the entire image - they should be centered around the updates
- if the updates are evenly distributed, increase the threshold
- set DCThreshold.t_default also to 0.1: this will truncate small updates at every activation layer.

If you still can't achieve a speedup, I'd be interested in taking a look at your implementation and data that you are using to analyse why the performance is lower than expected. If you are able to share that we me, of course. 
Contact me at mathias.parger@icg.tugraz.at

**Comment by IwakuraRein** [comment]
@dabeschte 
I added the code `print(self.sparsify(g_buffer)[1].float().mean())` and it said the sparsity is around 0.52.

However, I noticed a weird thing. After I added this line, there seemed an improvement in speed. I tested the time cost again with following codes:
```python
# ...
def forward(self, inp):
    torch.cuda.synchronize()
    starter, ender = torch.cuda.Event(enable_timing=True), torch.cuda.Event(enable_timing=True)
    starter.record()

    # print(self.sparsify(g_buffer)[1].float().mean())
    self.sparsify(g_buffer)[1].float().mean() # avoid the time cost of print function

    encodes = self.densify(self.network(self.sparsify(inp)))

    ender.record()
    torch.cuda.synchronize()
    print(starter.elapsed_time(ender))
    # ...
```

The time is about 4.3 - 6.7 ms. I got this result when there are other programs running on my GPU and it was better than the original 8.2 ms (tested when GPU was empty). Do you have any idea why simply adding a line could bring improvement?

EDIT:

After adding this line the output `encodes` are all zeros so this might be nothing but a bug. 

Anyway, if you want to reproduce, here are some of my input features: https://drive.google.com/file/d/1tl06AYAfCADbGUJ6LqwXLD1IiSMxKqVa/view?usp=sharing. The input is the concatenation of {\*.png, albedo\*.exr, shading_normal\*.exr}.

**Comment by dabeschte** [comment]
this is expected behaviour, not a bug. self.sparsify returns the difference between current and previous inputs. I assume that g_buffer is the same as inp here, in this case you essentially use the same input twice in a row, resulting in a zero difference between the two calls. what you do instead is save the result of the first call and use it again as input to the network.
e.g. 

```
# ...
def forward(self, inp):
    torch.cuda.synchronize()
    starter, ender = torch.cuda.Event(enable_timing=True), torch.cuda.Event(enable_timing=True)
    starter.record()

    # print(self.sparsify(g_buffer)[1].float().mean())
    inp_sparse = self.sparsify(g_buffer)[1].float().mean() # avoid the time cost of print function

    encodes = self.densify(self.network(inp_sparse))

    ender.record()
    torch.cuda.synchronize()
    print(starter.elapsed_time(ender))
    # ...
```

A density of 0.52 sounds quite high - especially if it is not all part of a big cluster, but instead spread around the image. I will take a quick look at the data and see what you could do to improve the sparsity.
